pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java

```diff
@@ -0,0 +1,512 @@
+            } else if (ECDSA.equals(privateKey.getAlgorithm())) {
+                dataKeyCipher = Cipher.getInstance(ECIES, BouncyCastleProvider.PROVIDER_NAME);
+            } else {
+                log.error("Unsupported key type {} for key {}.", privateKey.getAlgorithm(), keyName);
+                return false;
+            }
+            dataKeyCipher.init(Cipher.DECRYPT_MODE, privateKey);
+            dataKeyValue = dataKeyCipher.doFinal(encryptedDataKey);
+
+        } catch (IllegalBlockSizeException | BadPaddingException | NoSuchAlgorithmException | NoSuchProviderException
+                | NoSuchPaddingException | InvalidKeyException e) {
+            log.error("Failed to decrypt data key {} to decrypt messages {}", keyName, e.getMessage());
+            return false;
+        }
+        dataKey = new SecretKeySpec(dataKeyValue, "AES");
+        return true;
+    }
+
+    private ByteBuf decryptData(MessageMetadata msgMetadata, ByteBuf payload) {
+
+        // unpack iv and encrypted data
+        ByteString ivString = msgMetadata.getEncryptionParam();
+        ivString.copyTo(iv, 0);
+
+        // decrypt data
+        GCMParameterSpec gcmParams = new GCMParameterSpec(ivLen * 8, iv);
+        ByteBuf targetBuf = null;
+        try {
+            cipher.init(Cipher.DECRYPT_MODE, dataKey, gcmParams);
+
+            ByteBuffer sourceNioBuf = payload.nioBuffer(payload.readerIndex(), payload.readableBytes());
+
+            int maxLength = cipher.getOutputSize(payload.readableBytes());
+            targetBuf = PooledByteBufAllocator.DEFAULT.buffer(maxLength, maxLength);
+            ByteBuffer targetNioBuf = targetBuf.nioBuffer(0, maxLength);
+
+            int decryptedSize = cipher.doFinal(sourceNioBuf, targetNioBuf);
+            targetBuf.writerIndex(decryptedSize);
+
+        } catch (InvalidKeyException | InvalidAlgorithmParameterException | IllegalBlockSizeException
+                | BadPaddingException | ShortBufferException e) {
+            targetBuf.release();
+            targetBuf = null;
+            log.error("Failed to decrypt message {}", e.getMessage());
+        }
+
+        return targetBuf;
+    }
+
+    /*
+     * Decrypt the payload using the data key. Keys used to encrypt data key can be retrieved from msgMetadata
+     * 
+     * @param msgMetadata Message Metadata
+     * 
+     * @param payload Message which needs to be decrypted
+     * 
+     * @param keyReader KeyReader implementation to retrieve key value
+     * 
+     * @return decryptedData if success, null otherwise
+     */
+    public ByteBuf decrypt(MessageMetadata msgMetadata, ByteBuf payload, CryptoKeyReader keyReader) {
+
+        // If dataKey is present, attempt to decrypt using the existing key
+        if (dataKey != null) {
+            ByteBuf decryptedData = decryptData(msgMetadata, payload);
+            // If decryption succeeded, data is non null
+            if (decryptedData != null) {
```

Comment:
should we keep cache for `dataKey`?
eg. if we have two producers and each has different dataKey then here on every msg coming from different producer will not be able to decrypt data and requires to derive dataKey again.?